src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -299,13 +293,6 @@ internal TensorFlowTransform(IHostEnvironment env, Arguments args, IDataView inp
 
             if (args.ReTrain)
             {
-                // Shuffling fails because ShuffleTransform is not RowToRowMapper
-                // https://github.com/dotnet/machinelearning/issues/1106
-                //if(args.Shuffle)
-                //{
-                //    input = new ShuffleTransform(env, new ShuffleTransform.Arguments(), input);
```

Comment:
>input [](start = 22, length = 5)

frankly, I think the shuffle functionality could be preserved quite easily. You just need to make sure that `ShuffleTransform` is only applied at training time, and doesn't get applied after training.

So, don't do
`input = new Shuffle(input)`
but rather
`trainData = new Shuffle(input)`.

Reply:
Then again, I don't know how important it is to shuffle, so I'm fine either way.

---
In reply to: [224558825](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224558825) [](ancestors = 224558825)

Reply:
Not quite sure how helpful it is to remove vs keep it , esp. since tests for it are also commented out. 

(don't quite see the value of training component at this point with/without shuffle)

---
In reply to: [224558981](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224558981) [](ancestors = 224558981,224558825)

Reply:
@Pete, Shuffle is important as can be witnessed by the performance of the tests(shuffle vs. non-shuffle). The test dataset is small but on larger datasets performance gains are much better. I have moved the shuffling out of transform to give user more control over the shuffling.

I have done the way you have mentioned but I am currently not sure what the end user scenario will be for transforms like "Shuffle", "Filter" etc.. If this would be the same in Estimator/Pigsty world that would be great then.

@Abhishek, code was commented for tests because of error. Now its no longer commented out as I have fixed it.

---
In reply to: [224568214](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224568214) [](ancestors = 224568214,224558981,224558825)

Reply:
This is not what I meant. You can have the same logic (train on shuffle, then apply without shuffle) right here in this code, it doesn't have to be done by the caller.

---
In reply to: [224573294](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224573294) [](ancestors = 224573294,224568214,224558981,224558825)

Reply:
@zeahmed I am not aware of having mentioned anything like that, but I think I agree with @Zruty0 on the issue.

Reply:
Ahh I see. That's actually I did not intend to do inside TFTransform. The reason being is that Shuffle has its own parameters e.g. `ForceShuffleSeed` which I will have to expose as another parameter in `TFTransform.Arguments` in addition to `Shuffle` (or may be more depending on usage). But I am fine if you think that's ok to have those parameters as `TFTransform.Arguments`.

I will change it to what you said unless you agree with my thinking...:)

---
In reply to: [224594321](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224594321) [](ancestors = 224594321,224573294,224568214,224558981,224558825)

Reply:
Sorry wrong Pete...CodeFlow seems to have issue with tagging people.

---
In reply to: [224607610](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224607610) [](ancestors = 224607610)

Reply:
So, I would be OK with either:
- Having no 'shuffle' at all, performing shuffling in the test, but still don't make the training constructor public: use estimator to `Fit` the transform.
- Having Shuffle available without extra shuffling params and doing training as I outlined above.
- Same, but with additional params to Shuffle also available.

My preference would be to do 2, or 1, but not 3, unless it is obvious that having extra args to Shuffle is actually helpful

---
In reply to: [224617477](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224617477) [](ancestors = 224617477,224594321,224573294,224568214,224558981,224558825)

Reply:
hmm...1 or 2 requires `TensorFlowEstimator`  to be non-trivial estimator then. Currently, `TensorFlowEstimator` is trivial estimator. I will change the `TensorFlowEstimator` to non-trivial one and update the PR title and description to reflect changes.

---
In reply to: [224640105](https://github.com/dotnet/machinelearning/pull/1208#discussion_r224640105) [](ancestors = 224640105,224617477,224594321,224573294,224568214,224558981,224558825)